### PR TITLE
[tune] Fast Node Recovery

### PR DESCRIPTION
**Pull request by richardliaw**
## What do these changes do?

Pre-empting nodes will result in actors sometimes being "lost". This
ends up taking book-keeping space and also resources that aren't filled. 

This PR aims to address that.

## Linter

- [x] I've run `scripts/format.sh` to lint the changes in this PR.

### CODE REVIEW

File: python/ray/autoscaler/commands.py [hunk 1]
```diff
@@ -131,19 +131,21 @@ def kill_node(config_file, yes, override_cluster_name):
...
-            auth_config=config["auth"],
-            cluster_name=config["cluster_name"],
-            file_mounts=config["file_mounts"],
-            initialization_commands=[],
-            setup_commands=[],
-            runtime_hash="")
-
-        _exec(updater, "ray stop", False, False)
+        if hard:
+            provider.terminate_node(node)
```
Comment (hartikainen) [inline python/ray/autoscaler/commands.py]: Why is this way of killing the node not used by default? Is it because it can possibly block indefinitely?

File: python/ray/autoscaler/commands.py [hunk 2]
```diff
@@ -157,6 +159,13 @@ def kill_node(config_file, yes, override_cluster_name):
     return node_ip
 
 
+def monitor_cluster(cluster_config_file, override_cluster_name):
```
Comment (hartikainen) [inline python/ray/autoscaler/commands.py]: Nice!

File: python/ray/autoscaler/commands.py [hunk 3]
```diff
@@ -157,6 +159,13 @@ def kill_node(config_file, yes, override_cluster_name):
...
+    """Kills a random Raylet worker."""
+    cmd = "tail -n 100 -f /tmp/ray/session_*/logs/monitor*"
```
Comment (hartikainen) [inline python/ray/autoscaler/commands.py]: Could make the tail length an argument. Sometimes 100 lines doesn't really give you much information and you want to see logs further back.

File: python/ray/autoscaler/commands.py (hunk 2 above)
Comment (richardliaw) [reply python/ray/autoscaler/commands.py]: Should these go in different PRs?

File: python/ray/autoscaler/commands.py (hunk 1 above)
Comment (richardliaw) [reply python/ray/autoscaler/commands.py]: Because killing the raylet is faster for some previous fault tolerance tests - should this be a separate PR?

Comment (richardliaw) [reply python/ray/autoscaler/commands.py]: also cc @stephanie-wang

File: python/ray/autoscaler/commands.py (hunk 2 above)
Comment (richardliaw) [reply python/ray/autoscaler/commands.py]: Took it out!

File: python/ray/autoscaler/commands.py (hunk 3 above)
Comment (richardliaw) [reply python/ray/autoscaler/commands.py]: Done (#5082)

File: python/ray/autoscaler/autoscaler.py [hunk 4]
```diff
@@ -196,6 +196,7 @@ def prune(mapping):
                     "LoadMetrics: "
                     "Removed {} stale ip mappings: {} not in {}".format(
                         len(unwanted), unwanted, active_ips))
+            assert not unwanted.intersection(set(mapping))
```
Comment (hartikainen) [inline python/ray/autoscaler/autoscaler.py]:
```suggestion
            assert not (unwanted & set(mapping))
```

File: python/ray/autoscaler/aws/node_provider.py [hunk 5]
```diff
@@ -204,15 +208,30 @@ def create_node(self, node_config, tags, count):
...
-        conf.update({
-            "MinCount": 1,
-            "MaxCount": count,
-            "SubnetId": subnet_id,
-            "TagSpecifications": tag_specs
-        })
-        self.ec2.create_instances(**conf)
+
+        max_retries = 5
+        for attempt in range(1, max_retries + 1):
```
Comment (hartikainen) [inline python/ray/autoscaler/aws/node_provider.py]: Why is this needed?

File: python/ray/state.py [hunk 6]
```diff
@@ -382,8 +382,12 @@ def client_table(self):
             Information about the Ray clients in the cluster.
         """
         self._check_connected()
+        client_table = _parse_client_table(self.redis_client)
 
-        return _parse_client_table(self.redis_client)
+        for client in client_table:
+            client["alive"] = (client["EntryType"] !=
+                               gcs_utils.ClientTableData.DELETION)
```
Comment (hartikainen) [inline python/ray/state.py]: Are there possibly other `EntryType`s that should be considered as not-alive?

File: python/ray/autoscaler/aws/node_provider.py (hunk 5 above)
Comment (richardliaw) [reply python/ray/autoscaler/aws/node_provider.py]: This isn't directly related to the PR but occasionally I'll get an issue with the subnet not being compatible with `Security group X and subnet Y belong to different networks`. Since we randomly choose a subnet, I just made it retry a couple times.

File: python/ray/state.py (hunk 6 above)
Comment (richardliaw) [reply python/ray/state.py]: I think this is it - I'm using this as a reference (ray/state.py:cluster_resources):

```python
    def cluster_resources(self):
        """Get the current total cluster resources.

        Note that this information can grow stale as nodes are added to or
        removed from the cluster.

        Returns:
            A dictionary mapping resource name to the total quantity of that
                resource in the cluster.
        """
        self._check_connected()

        resources = defaultdict(int)
        clients = self.client_table()
        for client in clients:
            # Only count resources from latest entries of live clients.
            if client["EntryType"] != gcs_utils.ClientTableData.DELETION:
                for key, value in client["Resources"].items():
                    resources[key] += value
        return dict(resources)
```

**Comment by hartikainen** [comment]
This looks good to me. I'm just about to test this out, will report later how it goes.

@sudeepDasari have you had a chance to try this out? If so, how did it go?

**Comment by SudeepDasari** [comment]
@hartikainen I've been running this for the last day or so on 8 worker nodes and have seen 9 successful restarts thus far. Thank you so much, it seems to work!

File: python/ray/tune/ray_trial_executor.py [hunk 7]
```diff
@@ -288,7 +288,29 @@ def get_running_trials(self):
...
+        for node in nodes:
+            if node["alive"]:
+                ip_addresses.add(node["NodeManagerAddress"])
+        return ip_addresses
+
+    def get_current_trial_ips(self):
+        return set(t.node_ip for t in self.get_running_trials())
+
     def get_next_available_trial(self):
+        if not ray.worker._mode() == ray.worker.LOCAL_MODE:
```
Comment (hartikainen) [inline python/ray/tune/ray_trial_executor.py]:
```suggestion
        if ray.worker._mode() != ray.worker.LOCAL_MODE:
```

File: python/ray/tune/ray_trial_executor.py [hunk 8]
```diff
@@ -288,7 +288,29 @@ def get_running_trials(self):
 
         return list(self._running.values())
 
+    def get_node_ips(self):
```
Comment (hartikainen) [inline python/ray/tune/ray_trial_executor.py]:
```suggestion
    def get_alive_node_ips(self):
```

File: python/ray/tune/ray_trial_executor.py [hunk 9]
```diff
@@ -288,7 +288,29 @@ def get_running_trials(self):
...
+            if node["alive"]:
+                ip_addresses.add(node["NodeManagerAddress"])
+        return ip_addresses
+
+    def get_current_trial_ips(self):
+        return set(t.node_ip for t in self.get_running_trials())
+
     def get_next_available_trial(self):
+        if not ray.worker._mode() == ray.worker.LOCAL_MODE:
+            live_cluster_ips = self.get_node_ips()
```
Comment (hartikainen) [inline python/ray/tune/ray_trial_executor.py]:
```suggestion
            live_cluster_ips = self.get_alive_node_ips()
```

File: python/ray/tune/ray_trial_executor.py [hunk 10]
```diff
@@ -541,8 +563,13 @@ def restore(self, trial, checkpoint=None):
                 assert type(value) != Checkpoint, type(value)
                 trial.runner.restore_from_object.remote(value)
             else:
-                worker_ip = ray.get(trial.runner.current_ip.remote())
-                trial.sync_logger_to_new_location(worker_ip)
+                # This can be very slow - a better fix would
```
Comment (hartikainen) [inline python/ray/tune/ray_trial_executor.py]: Maybe add a TODO here and clarification when this would be very slow? Also might want to create an issue if this can cause problems.

File: python/ray/tune/trial_runner.py [hunk 11]
```diff
@@ -597,9 +597,19 @@ def _requeue_trial(self, trial):
 
         This does not notify the SearchAlgorithm because the function
         evaluation is still in progress.
+
         """
         self._scheduler_alg.on_trial_error(self, trial)
         self.trial_executor.set_status(trial, Trial.PENDING)
+
+        # Right now, this requeues the trial to the end of the queue. This is
```
Comment (hartikainen) [inline python/ray/tune/trial_runner.py]: This should also have a TODO and possibly an issue for fixing this?

**Review by hartikainen — APPROVED** [review]
Apart from the couple of comments, I think this looks good.
